File: sorting/selection_sort.cpp
#include <iostream>
#include <vector>

using namespace std;

template<typename T>
void selection_sort(vector<T> &A) {
  size_t pos = A.size() - 1;
  
  for (; pos > 0; pos--) {
    size_t max_index = pos;
    for (size_t current_index = 0; current_index < pos; current_index++) {
      if (A[current_index] > A[max_index]) {
        max_index = current_index;
      }
    }
    swap(A[pos], A[max_index]);
  }
}

template<typename T>
void reverse_selection_sort(vector<T> &A) {
  size_t pos = A.size() - 1;
  
  for (; pos > 0; pos--) {
    size_t max_index = pos;
    for (size_t current_index = 0; current_index < pos; current_index++) {
      if (A[current_index] < A[max_index]) {
        max_index = current_index;
      }
    }
    swap(A[pos], A[max_index]);
  }
}

int main() {
  // asc
  cout << "--------- asc ---------" << endl;
  vector<int> numbers = {5,4,3,2,1};
  selection_sort(numbers);
  for (size_t index = 0; index < numbers.size(); index++) {
    cout << numbers[index] << endl;
  }

  cout << endl;

  // desc
  cout << "--------- desc ---------" << endl;
  vector<int> reverse_numbers = {1,2,3,4,5};
  reverse_selection_sort(reverse_numbers);
  for (size_t index = 0; index < reverse_numbers.size(); index++) {
    cout << reverse_numbers[index] << endl;
  }
}

File: sorting/selection_sort.js
function selectionSort(arr) {
  let pos = arr.length - 1

  for (; pos > 0; pos--) {
    let max_index = pos
    for (let index = 0; index < pos; index++) {
      if (arr[index] > arr[max_index]) {
        max_index = index
      }
    }
    const temp = arr[pos]
    arr[pos] = arr[max_index]
    arr[max_index] = temp
  }
}

function reverseSelectionSort(arr) {
  let pos = arr.length - 1

  for (; pos > 0; pos--) {
    let max_index = pos
    for (let index = 0; index < pos; index++) {
      if (arr[index] < arr[max_index]) {
        max_index = index
      }
    }
    const temp = arr[pos]
    arr[pos] = arr[max_index]
    arr[max_index] = temp
  }
}

const numbers = Array(1000).fill(undefined).map((_, index) => {
  return 1000-index
})
console.time('built-in sort')
numbers.sort((a, b) => {
  return a-b
}) // built-in sort: 0.207ms
console.timeEnd('built-in sort')

const numbers2 = Array(1000).fill(undefined).map((_, index) => {
  return 1000-index
})
console.time('selection sort')
selectionSort(numbers2) // selection sort: 4.850ms
console.timeEnd('selection sort')

const reverseNumbers = [1,2,3,4,5]
reverseSelectionSort(numbers)
console.log(numbers)
